core/tx_pool.go

```diff
@@ -610,6 +612,38 @@ func (pool *TxPool) Pending(enforceTips bool) map[common.Address]types.Transacti
 	return pending
 }
 
+// PendingWithPredicates checks all currently processable transactions, grouped by origin
+// account and sorted by nonce. For each origin account, all transactions after and including
+// the first failed predicate are removed from the pool. The returned transaction set is a
+// copy and can be freely modified by calling code.
+//
+// The enforcePredicates parameter can be used to enforce precompile predicates on any pending
+// transactions that reference a precompile in its access list and remove any transactions
+// from the pool that fail to meet the predicate
+func (pool *TxPool) PendingWithPredicates(enforcePredicates bool, rules params.Rules, snowCtx *snow.Context, proposerVMBlockCtx *block.Context) map[common.Address]types.Transactions {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	pending := make(map[common.Address]types.Transactions)
+	for addr, list := range pool.pending {
+		txs := list.Flatten()
+
+		// If the miner requests predicate enforcement, we remove all transactions after and including the index of the first failed predicate
+		if enforcePredicates && !pool.locals.contains(addr) {
```

Comment:
We cannot exempt locals here, this would lead to building an invalid block.

Reply:
We should also enforce tips here as well as enforcing the predicates rather than only enforcing the predicates.

For `enforceTips` we can exempt locals because the transaction is not invalid it is just paying a tip below what the miner has stated as its minimum tip for inclusion in the block.

Reply:
Fixed.

Reply:
Fixed.